public/index.ts

```diff
@@ -996,94 +894,49 @@ function onVolumeCreated(volume: Volume, isTimeSeries = false, frameNumber = 0)
     view3D.setVolumeRotation(myState.volume, alignTransform.rotation.toArray());
   }
 
-  updateTimeUI(myState.volume);
+  updateTimeUI();
   updateZSliceUI(myState.volume);
   showChannelUI(myState.volume);
 }
 
-function copyVolumeToVolume(src, dest) {
-  // assumes volumes already have identical dimensions!!!
-
-  // grab references to data from each channel and put it in myState.volume
-  for (let i = 0; i < src.num_channels; ++i) {
-    // dest.channels[i].imgData.data.set(src.channels[i].imgData.data);
-    // dest.channels[i].volumeData.set(src.channels[i].volumeData);
-    // dest.channels[i].lut.set(src.channels[i].lut);
-    dest.channels[i].imgData = {
-      data: src.channels[i].imgData.data,
-      width: src.channels[i].imgData.width,
-      height: src.channels[i].imgData.height,
-    };
-    dest.channels[i].volumeData = src.channels[i].volumeData;
-    dest.channels[i].lut = src.channels[i].lut;
-
-    // danger: not a copy!
-    dest.channels[i].histogram = src.channels[i].histogram;
-    dest.channels[i].dataTexture = src.channels[i].dataTexture;
-    dest.channels[i].lutTexture = src.channels[i].lutTexture;
-    dest.channels[i].loaded = true;
-  }
-}
-
-function updateViewForNewVolume() {
-  view3D.onVolumeData(myState.volume, [0, 1, 2]);
-  myState.volume.setIsLoaded(true);
-
-  if (myState.isPT) {
-    view3D.updateActiveChannels(myState.volume);
-  } else {
-    view3D.updateLuts(myState.volume);
-  }
-
-  for (let i = 0; i < myState.volume.imageInfo.numChannels; ++i) {
-    view3D.updateIsosurface(myState.volume, i, myState.channelGui[i].isovalue);
-  }
-}
-
 function playTimeSeries(onNewFrameCallback: () => void) {
-  clearInterval(myState.timerId);
+  window.clearTimeout(myState.timerId);
+  myState.isPlaying = true;
 
   const loadNextFrame = () => {
-    let nextFrame = myState.currentFrame + 1;
-    if (nextFrame >= myState.totalFrames) {
-      nextFrame = 0;
-    }
-    const nextFrameVolume = myState.timeSeriesVolumes[nextFrame];
-
-    copyVolumeToVolume(nextFrameVolume, myState.volume);
-    updateViewForNewVolume();
-    myState.currentFrame = nextFrame;
-    onNewFrameCallback();
+    myState.lastFrameTime = Date.now();
+    const nextFrame = (myState.currentFrame + 1) % getNumberOfTimesteps();
+
+    // TODO would be real nice if this were an `await`-able promise instead...
+    view3D.setTime(myState.volume, nextFrame, (vol) => {
+      if (vol.isLoaded()) {
+        myState.currentFrame = nextFrame;
+        onNewFrameCallback();
+
+        if (myState.isPlaying) {
+          const timeLoading = Date.now() - myState.lastFrameTime;
+          myState.timerId = window.setTimeout(loadNextFrame, PLAYBACK_INTERVAL - timeLoading);
+        }
```

Comment:
Should this check happen before setting the state (or even before calling `view3d.setTime`)? If playback is stopped, you might see the frame jump suddenly.

Reply:
The new recursive structure of this function is a bit harder to reason about (I had to think about this comment for a second before figuring out the answer too)... The callback runs after `view3D.setTime` has completed loading in new frame data, so the part outside the check is dealing with that, and the part inside it is about whether we're still playing and should do it again.

Zooming out: while playback is happening, this recursive loop is either waiting on a timeout or loading new data. If we're waiting on a timeout, `pause` will cancel that timeout, so there's no need to check `myState.isPlaying` in the outer scope of this function. There's no meaningful way to cancel data loading at the moment (so the frame may jump by one after the user has hit pause), but we can prevent another timeout from being fired in the first place. That's what the check is doing.